BREAKING CHANGE: update `HybridComposition` methods to use `Issue` type for hints

duckki opened the pull request:
### Background

We are anticipating our new Rust composition will be a drop-in replacement of HybridComposer. But, currently, `HybridComposition` trait is using the JS-specific `CompositionHint` type for hints (in both `MergeResult` and `SatisfiabilityResult`).

The problem is the Rust composition has to emulate the JS data structures as it currently stands. It will be much easier if the trait methods use the `Issue` type in the first place.

### Proposed Changes

This PR changes `HybridComposition` trait to use the `Issue` type directly, instead of JS `CompositionHint`. This is intended to be a pure refactoring. `CompositionHint` values from `HybridComposition` are immediately converted to `Issue` type anyways. We can use `Issue` type for API and do the conversion in the fedutils-build-plugin's implementation with no effective changes.

* changed `MergeResult` to use `Issue` type.
  * This is only used for `experimental_` methods. So, this is not a breaking change.
* moved `MergeResult` from `javascript` module to `composition` module, since it does not concern JS serialization.
* changed `validate_satisfiability` trait method to return `Vec<Issue>`, instead of `SatisfiabilityResult`.
  * This is technically a breaking change, but it shouldn't be used outside of this apollo-composition & fbp and not visible to customers.
  * If we want to keep the API in the main branch, we can target `next` branch, instead.
  * Note: The method previously returned `Result<SatisfiabilityResult, Issue>`, which really had 3 cases -- success with optional JS hints, satisfiability error with a list of JS errors, or deno error with one `Issue`. This PR conflates the latter two as `Err(Vec<Issue>)`. I think that's ok since the downstream usage of this method does not really distinguish those cases.

Companion PR on fbp repo: https://github.com/mdg-private/fedutils-build-plugins/pull/151

duckki committed:
updated HybridComposition::experimental_* interface to use `Vec<Issue>` for hints

dariuszkuc commented on apollo-composition/src/lib.rs:
```diff
@@ -161,12 +159,13 @@ pub trait HybridComposition {
         mut self,
         subgraph_definitions: Vec<SubgraphDefinition>,
-    ) -> Result<PluginResult, Vec<Issue>>
+    ) -> Result<(String, Vec<Issue>), Vec<Issue>>
```
IMHO returning proper type is clearer than a tuple, i.e. what is the meaning behind `String` and `Vec<Issue>` that are being returned?

duckki committed:
addressed review comment

- kept using PluginResult and MergeResult structs
- moved MergeResult to `plugin_result` mod, since it has nothing to do with JS serialization.

duckki committed:
reverted some unnecessary changes

duckki committed:
change the type signature of `HybridComposition::validate_satisfiability`

- This simplifies apollo-composition and future Rust composition implementation.
- The previous `satisfiability_result_into_issues` function will be moved to fedutils-build-plugins.
- This breaks API, but it's only used internally between federation-rs and fedutils-build-plugins.

duckki committed:
updated comments

duckki committed:
clippy fix

duckki committed:
Update apollo-composition/src/lib.rs

Accepted reviewer suggestion

Co-authored-by: Dariusz Kuc <9501705+dariuszkuc@users.noreply.github.com>

duckki committed:
refactor: move MergeResult to `composition` module.

duckki commented:
I created a new PR targeting the `next` branch: https://github.com/apollographql/federation-rs/pull/651
Closing.
